src/NBitcoin/ProvenBlockHeader.cs

```diff
@@ -80,19 +87,34 @@ public ProvenBlockHeader(PosBlock block)
             this.Version = block.Header.Version;
 
             // Set additional properties.
+            this.MerkleProofSize = null;
+            this.CoinstakeSize = null;
+            this.SignatureSize = null;
+
             this.signature = block.BlockSignature;
             this.coinstake = block.Transactions[1];
-
             this.merkleProof = new MerkleBlock(block, new[] { this.coinstake.GetHash() }).PartialMerkleTree;
         }
 
         /// <inheritdoc />
         public override void ReadWrite(BitcoinStream stream)
         {
             base.ReadWrite(stream);
+            long baseByteSize = GetProcessedBytes(stream);
+
             stream.ReadWrite(ref this.merkleProof);
+            this.MerkleProofSize = GetProcessedBytes(stream) - baseByteSize;
+
             stream.ReadWrite(ref this.signature);
+            this.SignatureSize = GetProcessedBytes(stream) - baseByteSize - this.MerkleProofSize;
+
             stream.ReadWrite(ref this.coinstake);
+            this.CoinstakeSize = GetProcessedBytes(stream) - baseByteSize - this.MerkleProofSize - this.SignatureSize;
+        }
+
+        private static long GetProcessedBytes(BitcoinStream stream)
```

Comment:
why static?

Reply:
also why not to the stream class itself?

Reply:
Static because it is a helper function and it is not dealing with the instance itself. I don't really mind either way, it just seems logical for functions of this type